src/mlpack/methods/hoeffding_trees/hoeffding_tree_impl.hpp

```diff
@@ -601,32 +533,35 @@ void HoeffdingTree<
          const arma::Row<size_t>& labels,
          const size_t numClasses,
          const bool batchTraining,
-         const double successProbability,
-         const size_t maxSamples)
+         const std::optional<double> successProbability,
+         const std::optional<size_t> maxSamples,
+         const std::optional<size_t> checkInterval)
 {
-  Train(data, info, labels, numClasses, batchTraining, successProbability,
-      maxSamples, this->checkInterval, this->minSamples);
-}
 
-template<typename FitnessFunction,
-         template<typename> class NumericSplitType,
-         template<typename> class CategoricalSplitType>
-template<typename MatType>
-void HoeffdingTree<
-    FitnessFunction,
-    NumericSplitType,
-    CategoricalSplitType
->::Train(const MatType& data,
-         const data::DatasetInfo& info,
-         const arma::Row<size_t>& labels,
-         const size_t numClasses,
-         const bool batchTraining,
-         const double successProbability,
-         const size_t maxSamples,
-         const size_t checkInterval)
-{
-  Train(data, info, labels, numClasses, batchTraining, successProbability,
-      maxSamples, checkInterval, this->minSamples);
+  if (successProbability.has_value())
+  {
+    Train(data, info, labels, numClasses, batchTraining,
+        successProbability.value(), this->maxSamples, this->checkInterval,
+        this->minSamples);
+  }
+  else if (successProbability.has_value() && maxSamples.has_value())
+  {
+    Train(data, info, labels, numClasses, batchTraining,
+        successProbability.value(), maxSamples.value(), this->checkInterval,
+        this->minSamples);
+  }
+  else if (successProbability.has_value() && maxSamples.has_value() &&
+      checkInterval.has_value())
+  {
+    Train(data, info, labels, numClasses, batchTraining,
+        successProbability.value(), maxSamples.value(), checkInterval.value(),
+        this->minSamples);
+  }
+  else
+  {
+    Train(data, info, labels, numClasses, batchTraining, this->successProbability,
+        this->maxSamples, this->checkInterval, this->minSamples);
+  }
```

Comment:
Oh, also, I realized the logic here is not correct and that is why the tests are failing: suppose you pass both `successProbability` and `maxSamples`... but the first `if` (`successProbability.has_value()`) will be taken, and this only passes the given `successProbability` but not the given `maxSamples`.  The suggestion I posted above should fix that, but there are other ways to solve the problem too if you prefer.

Reply:
Oh yes indeed it is a logic error, it will go to the first one that is validating the if condition, sorry I missed that.
I see the logic with the ternary operator, it should make things much easier.